Move `application` handling from `record` to `config.load`.

zwn opened the pull request:
With this change any module (including the application) can be replayed
using generic `replay` facility (one step closer to unified launcher).

Fixes #251

m3d commented:
Is this code backwards compatible (i.e. if I want to replay old `osgar.go` logfiles)?

```
python -m osgar.replay --module app go-kloubak2-subt-estop-lora-191007_153520.log
original args: b"['/home/robot/git/osgar/osgar/go.py', 'run', 'config/kloubak2-subt-estop-lora.json', '--dist', '10', '--timeout', '120', '--note', 'via sh script']"
stream names:
  app.desired_speed
  app.pose2d
  app.artf_xyz
  app.pose3d
  detector.artf
  detector.dropped
  kloubak.can
  kloubak.encoders
  kloubak.emergency_stop
  kloubak.pose2d
  can.can
  can.raw
  serial.raw
  lidar.raw
  lidar.scan
  lidar_tcp.raw
  lidar_back.raw
  lidar_back.scan
  lidar_back_tcp.raw
  camera.raw
  camera_back.raw
  imu.orientation
  imu.rotation
  imu_serial.raw
  estop.raw
  estop.emergency_stop
  estop_serial.raw
  lora.raw
  lora.cmd
  lora_serial.raw
inputs: ['encoders', 'scan', 'scan_back', 'emergency_stop', 'pose2d', 'buttons', 'voltage', 'rot']
outputs: ['desired_speed', 'pose2d', 'artf_xyz', 'pose3d']
Traceback (most recent call last):
  File "C:\Tool\WinPython-64bit-3.6.2.0Qt5\python-3.6.2.amd64\lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "C:\Tool\WinPython-64bit-3.6.2.0Qt5\python-3.6.2.amd64\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "c:\admin\_private\osgar\osgar\replay.py", line 80, in <module>
    module_instance = replay(args)
  File "c:\admin\_private\osgar\osgar\replay.py", line 61, in replay
    assert application is not None
AssertionError
```

zwn commented:
No. The handling of application-related functionality has been moved from `Recorder` to `config.load`. The code now expects that the driver for the module `app` is the actual module, ie. `osgar.go:Go`. Without this information `osgar.replay` does not know what app is supposed to be. Do you thing such a backward compatibility is really needed?

We _could_ add some cmdline options to `osgar.replay` like `--application=<something>`. I just felt that application logs are such a short lived thing it does not matter (the log replay only works until you change the application code - unlike for example log of camera images).

Oh, I see I forgot some code in `replay` that still works with application, which should not be there anymore so we will need some updates anyway.

m3d commented:
There used to be also option to run recorded data on "different application" - say you use "Go 1 meter" to scan the environment and then try some "explore" because you recorded some interesting corner. I guess that force another configuration may still work??? I will test it on Eduro tonight (35%) and then I would merge it.

zwn commented:
Yes - you can use another configuration. Let me know how you feel about the ergonomics, we can also tweak the details later if required.

m3d commented:
```
root@voyage:/home/robot/git/osgar# python -m osgar.go run config/eduro-subt.json  --note "test before app config load change" -d 0.2
Namespace(command='run', config=['config/eduro-subt.json'], dist=0.2, note='test before app config load change', speed=None, timeout=10)
Traceback (most recent call last):
  File "/root/miniconda3/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/root/miniconda3/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/robot/git/osgar/osgar/go.py", line 93, in <module>
    config = config_load(args.config, application=Go)
  File "/home/robot/git/osgar/osgar/lib/config.py", line 44, in load
    with open(filename) as f:
TypeError: expected str, bytes or os.PathLike object, not list
```
was it working on your robot?

m3d commented:
I hacked it with
```
root@voyage:/home/robot/git/osgar# git diff
diff --git a/osgar/lib/config.py b/osgar/lib/config.py
index 4f962c6..40e8745 100644
--- a/osgar/lib/config.py
+++ b/osgar/lib/config.py
@@ -41,7 +41,8 @@ def _application2import(application):
 def load(*filenames, application=None):
     ret = {}
     for filename in filenames:
-        with open(filename) as f:
+        print(filename)
+        with open(filename[0]) as f:
             data = json.load(f)
             assert 'version' in data, data
             assert data['version'] in SUPPORTED_VERSIONS, data['version']
```
because the `filename` was `['config/eduro-subt.json']` ... but I am not sure about other consequences.

m3d commented:
Replay then looks like working and in log is stored `{'app': {'driver': 'osgar.go:Go'` or `{'app': {'driver': 'osgar.turn:Turn'`

zwn commented:
Sorry about that bug :worried: . I changed the signature of the `config.load` function and didn't fix all the usages properly afterwards. It is really hard to find and fix all usages of `config.load` due to the fact they are sometimes renamed on import to `config_load` and sometimes not. That is mostly why I renamed it in #256. And why I proposed #255. But it should be fixed now.

m3d commented:
OK, thanks - can you rebase it and prepare for merge? I would test it on Kloubak tonight and if it works then I would merge it, OK?

m3d commented:
maybe merge this after #256 and I will ask for fix in case of failure, OK?

zwn commented:
I have rebased it on current master. I have also fixed `osgar.replay.replay` to the new usage of `config_load`. With this adjustment it allows for backward compatibility using different config file and replacing `application` driver with the supplied driver. However, this isn't directly exposed in anyway (but might be helpful in one-off scripts). Is it ok like this?

Would now be a good time to remove custom replay code from all the "main" modules? It used to be there to allow application to be replayed which can now be handled by the generic replay facility.

zwn commented:
Let me correct myself - it is used in launcher. So when replaying through launcher and custom config file is specified, the supplied application is used.

zwn commented:
Ok, back to the trees as they say :wink: . When using `replay.replay` and supplying custom config file, the application is used. If not supplying custom config, the argument `application` is ignored.

It is not nice, but it is working. I would keep it this way for this PR and refactor later (with the removal of the replay sections in general, on the move to one common launcher).

I am not aware of any further bugs in this PR (:tm:) :wink:
